doc/rf.in.man
.\" Originally generated by cmd.
.TH RF "1" "Monday July 2014" "rf 2014.06.07" "User Commands"
.SH NAME
rf \- Feed Reader
.SH SYNOPSIS
.B rf [OPTIONS] [FEED]
.SH EXAMPLE
.B rf linux
.SH DESCRIPTION
read feed is a command that reads feeds from any source. rf uses the feed of a site to manage it with a command line interface.
.PP
Mandatory arguments for long options are mandatory for short options too.
.TP
\fB\-l\fR, \fB\-\-list\fR
list the available feeds
.TP
\fB\-t\fR, \fB\-\-title\fR[=\fIITEMS\fR]\fR
set title mode
.TP
\fB\-d\fR, \fB\-\-description\fR=\fIITEM\fR
set description mode
.TP
\fB\-n\fR, \fB\-\-link\fR=\fIITEM\fR
set link mode
.TP
\fB\-c\fR, \fB\-\-cache\fR=\fIMINUTES\fR
set cache time
.TP
\fB\-v\fR, \fB\-\-version\fR
show the program version
.TP
\fB\-v\fR, \fB\-\-variable\fR=\fIVARIABLE\fR
show a program variable
.TP
\fB\-h\fR, \fB\-\-help\fR
show the program help
.TP
\fB\-m\fR, \fB\-\-man\fR
show the program manual
.TP
On no config creates ~/.rf directory and adds the default feeds to the ~/.rf/feeds file. With no option, use title mode. With no items, show all. With no cache time, use 30 minutes.
.TP
The available --variable VARIABLES are: package, name, version, description, explanation, license, homepage, author, email, social, blog, usage, options, examples, extrahelp, extranotes, month, year, day and timestamp.
.TP
Execute 'bash -c "man <(rf --man)"' to see the runtime manpage.
.SH NOTES
For more information, check documentation.
.SH HOMEPAGE
Main project page at \fBhttp://code.google.com/p/codemiscs\fR.
.SH AUTHOR
Programmed by \fBJuan Manuel Borges Caño\fR.
.SH EMAIL
Thank, Contact, Follow through \fBjuanmabcmail@gmail.com\fR.
.SH SOCIAL
Interact with \fBhttp://plus.google.com/+JuanManuelBorgesCaño\fR
.SH SUPPORT
Support through Paypal, Patreon, Flattr, Western Union, BitCoin, ...
.SH COPYRIGHT
Copyright \(co \fB2014 Juan Manuel Borges Caño\fR.
.PP
This software is licensed and under the terms of the GPLv3+ License.
There is NO WARRANTY, to the extent permitted by law.


NEWS
2014-07-03 Juan Manuel Borges Caño <juanmabcblog.blogspot.com, juanmabcmail@gmail.com> 
	* New Commands: tonize, status.
	* Improved Commands: colorize, tw.
	* Improved Website: better looking design with more easy access to stuff that matters, and more content like NEWS & ChangeLog.
	* Better Design: colorize understands pattern and line matches, removes reverse duplication, tw adapts way faster to online engines changes. Website makes core usage of templates & is better metatagged.
	* Better Documentation: continues improving effortlessly through documentation automatic syndication, automatic tonize & status man pages.
	* Better Distribution: more automated generation of documentation and wiki.
	* Better Motivation: proved benefits sweetens development, nicer web with a lot of content exploit.

2014-03-16 Juan Manuel Borges Caño <juanmabcblog.blogspot.com, juanmabcmail@gmail.com> 
	* Proved Benefits: previous features consolidate actual ones.
	* Better Motivation: proved benefits sweetens development.
	* Improved Commands: cmd (core), typewrite, colorize.
	* Improved Documentation: more rich content with media.
	* Improved Distribution: more automated generation of binary packages.
	* Better Design: typewrite can scroll, and pass through non printable characters, colorize understands more keywords.
	* Better Website: thanks to better documentation and better distribution.

2014-02-10 Juan Manuel Borges Caño <juanmabcblog.blogspot.com, juanmabcmail@gmail.com> 
	* Improved Commands: cmd (core), tw, show.
	* Better Website: Continues improving effortlessly through documentation automatic syndication and smart meta distribution. Demonstrations of every command.
	* Better Design: The cmd improvements benefit instantly all commands coding, look and usage experiences. Show plugin design is more modulable and easily expandable.
	* Better Documentation: Continues improving effortlessly through documentation automatic syndication. Code parchments got everywhere where matters.
	* Better Distribution: Introducing smart meta distribution technology (currently targeted to open build service, build.opensuse.org instance, supporting fedora, ubuntu, suse, debian, centos, arch and autotools).
	* Proved Benefits: previous releases features consolidate into real value.
	* Better Motivation: copyrighted for 2014.

2013-12-30 Juan Manuel Borges Caño <juanmabcblog.blogspot.com, juanmabcmail@gmail.com> 
	* New Commands, Improved Commands, Better Design, Better Documentation, Better Motivation.

2013-12-19 codemiscs <code.google.com/p/codemiscs>
	* New Commands, Better Documentation, Better Website, Better Distribution.

2013-11-28 codemiscs <code.google.com/p/codemiscs>
	* Independent Projects Merger.


cmds/rf/rf.in.sh
#! /usr/bin/env bash

#  _________________________________________________________________________
# /\                                                                        \
# \_|       ___        __                            _   _                  |
#   |      |_ _|_ __  / _| ___  _ __ _ __ ___   __ _| |_(_) ___  _ __       |
#   |       | || '_ \| |_ / _ \| '__| '_ ` _ \ / _` | __| |/ _ \| '_ \      |
#   |       | || | | |  _| (_) | |  | | | | | | (_| | |_| | (_) | | | |     |
#   |      |___|_| |_|_|  \___/|_|  |_| |_| |_|\__,_|\__|_|\___/|_| |_|     |
#   |                                                                       |
#   |        rf (Read Feed): Feed Reader, read feeds from any source        |
#   |            Copyright (C) 2012 - 2014 Juan Manuel Borges Caño          |
#   |       The desire for a fast, command line, non interactive, rich      |
#   |               news feed reader encouraged this command.               |
#   |                    _     _                                            |
#   |                   | |   (_) ___ ___ _ __  ___  ___                    |
#   |                   | |   | |/ __/ _ \ '_ \/ __|/ _ \                   |
#   |                   | |___| | (_|  __/ | | \__ \  __/                   |
#   |                   |_____|_|\___\___|_| |_|___/\___|                   |
#   |                                                                       |
#   |  This program is free software: you can redistribute it and/or modify |
#   |  it under the terms of the GNU General Public License as published by |
#   |   the Free Software Foundation, either version 3 of the License, or   |
#   |                  (at your option) any later version.                  |
#   |                                                                       |
#   |    This program is distributed in the hope that it will be useful,    |
#   |     but WITHOUT ANY WARRANTY; without even the implied warranty of    |
#   |     MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the     |
#   |              GNU General Public License for more details.             |
#   |                                                                       |
#   |   You should have received a copy of the GNU General Public License   |
#   | along with this program.  If not, see <http://www.gnu.org/licenses/>. |
#   |   ____________________________________________________________________|_
#    \_/______________________________________________________________________/

declare -Ag rf_rss_feeds
declare -Ag rf_atom_feeds

# The setup helper.
function rf_setup
{
	[[ -z "$1" ]] && cmd_error "feed not specified"

	rf_feedurl="${rf_rss_feeds[$1]}"
	if [[ -z "$rf_feedurl" ]]
	then
		rf_feedurl="${rf_atom_feeds[$1]}"
		rf_type="atom"
	else
		rf_type="rss"
	fi
	[[ -z "$rf_feedurl" ]] && cmd_error "unknown feed"

	if [[ -f "$HOME/.rf/$1.rfd" ]] && (( $(date "+%s") - $(stat -c "%Y" "$HOME/.rf/$1.rfd") < 60 * "$rf_time" ))
	then
		rf_data="$(cat "$HOME/.rf/$1.rfd")"
	else
		rf_data="$(curl -s --connect-timeout 300 --max-time 300 -A "Mozilla/5.0" -L "$rf_feedurl")" && if [[ -d "$HOME/.rf" ]]; then printf "%s\n" "$rf_data" > "$HOME/.rf/$1.rfd"; fi
	fi

	rf_feedns="$(printf "%s\n" "$rf_data" | grep -m 1 -o 'xmlns="[^"]*"' | cut -d'"' -f2)"
}

# The title helper.
function rf_mode_title
{
	if [[ -z "$rf_feedns" ]]
	then
		if [[ "$rf_type" == "rss" ]]
		then
			rf_output="$(printf "%s\n" "$rf_data" | xmlstarlet sel -t -m "//item" -v "$rf_mode" -n)"
		else
			rf_output="$(printf "%s\n" "$rf_data" | xmlstarlet sel -t -m "//entry" -v "$rf_mode" -n)"
		fi
	else
		if [[ "$rf_type" == "rss" ]]
		then
			rf_output="$(printf "%s\n" "$rf_data" | xmlstarlet sel -N my="$rf_feedns" -t -m "//my:item" -v "my:$rf_mode" -n)"
		else
			rf_output="$(printf "%s\n" "$rf_data" | xmlstarlet sel -N my="$rf_feedns" -t -m "//my:entry" -v "my:$rf_mode" -n)"
		fi
	fi
	rf_i=1
	while (( $rf_items < 0 || $rf_i <= $rf_items )) && read -r rf_line 
	do
		printf "[$rf_i] %s\n" "$rf_line" | lynx -assume-charset="UTF-8" -dump -nolist -stdin | lynx -assume-charset="UTF-8" -dump -nolist -stdin | sed -e 's/^[[:space:]][[:space:]][[:space:]]//' -e '/^$/d' | fmt --width 2048
		(( rf_i++ ))
	done < <(printf "%s\n" "$rf_output")
}

# The description helper.
function rf_mode_description
{
	if [[ -z "$rf_feedns" ]]
	then
		if [[ "$rf_type" == "rss" ]]
		then
			rf_output="$(printf "%s\n" "$rf_data" | xmlstarlet sel -t -m "//item[$rf_item]" -v "$rf_mode")"
		else
			rf_output="$(printf "%s\n" "$rf_data" | xmlstarlet sel -t -m "//entry[$rf_item]" -v "content")"
		fi
	else
		if [[ "$rf_type" == "rss" ]]
		then
			rf_output="$(printf "%s\n" "$rf_data" | xmlstarlet sel -N my="$rf_feedns" -t -m "//my:item[$rf_item]" -v "my:$rf_mode")"
		else
			rf_output="$(printf "%s\n" "$rf_data" | xmlstarlet sel -N my="$rf_feedns" -t -m "//my:entry[$rf_item]" -v "my:content")"
		fi
	fi
	printf "%s\n" "$rf_output" | lynx -assume-charset="UTF-8" -dump -nolist -stdin | lynx -assume-charset="UTF-8" -dump -nolist -stdin | sed -e 's/^[[:space:]][[:space:]][[:space:]]//' -e '/^$/d' | fmt --width 2048
}

# The link helper.
function rf_mode_link
{
	if [[ -z "$rf_feedns" ]]
	then
		if [[ "$rf_type" == "rss" ]]
		then
			rf_output="$(printf "%s\n" "$rf_data" | xmlstarlet sel -t -m "//item[$rf_item]" -v "$rf_mode")"
		else
			rf_output="$(printf "%s\n" "$rf_data" | xmlstarlet sel -t -m "//entry[$rf_item]" -v "$rf_mode/@href")"
		fi
	else
		if [[ "$rf_type" == "rss" ]]
		then
			rf_output="$(printf "%s\n" "$rf_data" | xmlstarlet sel -N my="$rf_feedns" -t -m "//my:item[$rf_item]" -v "my:$rf_mode")"
		else
			rf_output="$(printf "%s\n" "$rf_data" | xmlstarlet sel -N my="$rf_feedns" -t -m "//my:entry[$rf_item]" -v "my:$rf_mode/@href")"
		fi
	fi
	printf "%s\n" "$rf_output"
}

# The --list option.
function rf_list
{
	for rf_feed in "${!rf_rss_feeds[@]}" "${!rf_atom_feeds[@]}"
	do
		printf "%s\n" "$rf_feed"
	done | sort
	exit 0
}

# The --title option.
function rf_title
{
	rf_mode="title"	
	if [[ -n "$1" ]]
	then
		rf_items="$1"
	fi
}

# The --description option.
function rf_description
{
	rf_mode="description"	
	rf_item="$1"
}

# The --link option.
function rf_link
{
	rf_mode="link"
	rf_item="$1"
}

# The --cache option.
function rf_cache
{
	rf_time="$1"
}

# ... and rf, the program itself.

# The cmd init function
function rf_init
{
	[[ ! -d "$HOME/.rf" ]] && mkdir "$HOME/.rf"
	[[ ! -f "$HOME/.rf/feeds" ]] && cp "$cmd_datadir/feeds" "$HOME/.rf/feeds"
	source "$HOME/.rf/feeds"
	rf_time="30"
	rf_mode="title"
	rf_items="-1"
	rf_item="0"
	rf_type=""
}

# The cmd main function
function rf_main
{
	rf_setup "$@"

	case "$rf_mode" in
		"title") rf_mode_title ;;
		"description") rf_mode_description ;;
		"link") rf_mode_link ;;
	esac
}

# The cmd fields
cmd_package="[@]pkg[@]"
cmd="rf"
cmd_name="read feed"
cmd_version="[@]pkgversion[@]"
cmd_description="Feed Reader"
cmd_explanation="read feed is a command that reads feeds from any source. rf uses the feed of a site to manage it with a command line interface."
cmd_license="[@]pkglicense[@]"
cmd_homepage="[@]pkghomepage[@]"
cmd_author="[@]pkgauthor[@]"
cmd_email="[@]pkgemail[@]"
cmd_social="[@]pkgsocial[@]"
cmd_blog="[@]pkgblog[@]"
cmd_usage="$cmd [OPTIONS] [FEED]"
cmd_options=("/l/list/list the available feeds/rf_list/" "/t::/title::/set title mode/rf_title/ITEMS/" "/d:/description:/set description mode/rf_description/ITEM/" "/n:/link:/set link mode/rf_link/ITEM/" "/c:/cache:/set cache time/rf_cache/MINUTES/")
cmd_extrahelp="On no config creates ~/.$cmd directory and adds the default feeds to the ~/.rf/feeds file. With no option, use title mode. With no items, show all. With no cache time, use 30 minutes."
cmd_examples=("$cmd linux")
cmd_extranotes="For more information, check documentation."
cmd_init="rf_init"
cmd_main="rf_main"

cmd_datadir="[@]pkgdatadir[@]/$cmd"

# The cmd environment
source "[@]pkgdatadir[@]/cmd.sh"
